fairmq/FairMQDevice.cxx

```diff
@@ -96,16 +145,43 @@ FairMQDevice::FairMQDevice(const fair::mq::tools::Version version)
 
```

Comment:
Can't we use delegating constructor here? Furthermore, owning and not owning the config member seems complicated, wouldn't this be a good case for `std::shared_pointer`?

```cpp
FairMQDevice::FairMQDevice()
    : FairMQDevice(std::make_shared<FairMQProgOptions>(), {0, 0, 0})
{
}

FairMQDevice::FairMQDevice(std::shared_ptr<FairMQProgOptions> config)
    : FairMQDevice(config, {0, 0, 0})
{
}

FairMQDevice::FairMQDevice(const fair::mq::tools::Version version)
    : FairMQDevice(std::make_shared<FairMQProgOptions>(), version)
{
}

FairMQDevice::FairMQDevice(std::shared_ptr<FairMQProgOptions> config, const fair::mq::tools::Version version)
    : fTransportFactory(nullptr)
    , fTransports()
    , fChannels()
    , fConfig(config)
    , fId()
    , fInitialValidationFinished(false)
    , fInitialValidationCondition()
    , fInitialValidationMutex()
    , fPortRangeMin(22000)
    , fPortRangeMax(32000)
    , fDefaultTransportType(fair::mq::Transport::DEFAULT)
    , fDataCallbacks(false)
    , fMsgInputs()
    , fMultipartInputs()
    , fMultitransportInputs()
    , fChannelRegistry()
    , fInputChannelKeys()
    , fMultitransportMutex()
    , fMultitransportProceed(false)
    , fVersion(version)
    , fRate(0.)
    , fLastTime(0)
    , fRawCmdLineArgs()
{
}
```

 What do you think?

Reply:
Delegating - yes, will change!

shared_ptr would be a API change, since any external config object would then have to be a shared_ptr, which are now plain objects in all cases. From my perspective this makes user write more code. With two pointers, as proposed (albeit a bit complex internally as you point out), the API stays simple and without break.